### DISCUSSION TITLE
Support for eslint v9

### DISCUSSION THREAD

**Discussion by pnodet** [vercel/next.js#54238] [Help] [answered]
### Summary

Eslint has introduced a new config files system called `flat config` and using `eslint.config.js`.

As per the [documentation:](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file-formats)

> Warning
> We are transitioning to a new config system in ESLint v9.0.0. The config system shared on this page is currently the default but will be deprecated in v9.0.0. You can opt-in to the new config system by following the instructions in the [documentation](https://eslint.org/docs/latest/use/configure/configuration-files-new).

I migrated my config to the flat system but nextjs doesn't picked it up when calling next lint.

From `next.js/packages/next/src/cli/next-lint.ts`

```js
    // Find user's .eslintrc file
    // See: https://eslint.org/docs/user-guide/configuring/configuration-files#configuration-file-formats
    const eslintrcFile =
      (await findUp(
        [
          '.eslintrc.js',
          '.eslintrc.cjs',
          '.eslintrc.yaml',
          '.eslintrc.yml',
          '.eslintrc.json',
          '.eslintrc',
        ],
        {
          cwd: baseDir,
        }
      )) ?? null
```

**Comment by pnodet** [comment]
Any update on this?

The flat config will be the de facto standard but I can't make it work when running next lint.

Saw a PR (#56181) but doesn't seem to be active...

**Comment by z0n** [comment]
ESLint 9 is now out and it's not compatible with `next lint` at all:
```
$ next lint
Invalid Options:
- Unknown options: useEslintrc, extensions, resolvePluginsRelativeTo, rulePaths, ignorePath, reportUnusedDisableDirectives
- 'extensions' has been removed.
- 'resolvePluginsRelativeTo' has been removed.
- 'ignorePath' has been removed.
- 'rulePaths' has been removed. Please define your rules using plugins.
- 'reportUnusedDisableDirectives' has been removed. Please use the 'overrideConfig.linterOptions.reportUnusedDisableDirectives' option instead.
```

If you get these errors, downgrade to ESLint 8.

**Reply by manuelbachl** [reply]
I get these exact messages right AFTER downgrading to ESLint 8...
Before I had the problem with the patch.

**Reply by kaungmyatlwin** [reply]
If you have `nextjs@15.3.0`, update to `eslint^9`. This solved for me.

**Comment by infodusha** [comment]
As a temporary solution you can skip eslint on build via config and run it manually. You have to reconfigure all the plugins to use flat config tho.

**Comment by dzienisz** [comment]
imho it should be updated in next 14

**Comment by developedbyumair** [comment]
downgrade  is not a solution

**Comment by poksme** [comment]
This will not make Next.js pickup your config file, but I wrote a guide on how to migrate to ESLint 9 until Next.js supports it : [Eslint 9 & Next.js — Setup Guide](https://medium.com/linotte-technology/eslint-9-next-js-935c2b6d0371)

Also for a TLDR; check my comment [here](https://github.com/vercel/next.js/issues/64409#issuecomment-2265632495)

**Comment by leerob** [comment, marked as answer]
ESLint 9 is supported as of the Next.js 15 RC 2. v15 stable release coming soon 🙏 

https://x.com/nextjs/status/1846276572812124565

**Reply by jdhiro** [reply]
I'm still not seeing any real details or examples of how we move to the new eslint.config.js format. I've tried many things but always get:

> Failed to patch ESLint because the calling module was not recognized.
> If you are using a newer ESLint version that may be unsupported, please create a GitHub issue:
> https://github.com/microsoft/rushstack/issues

**Reply by paschun** [reply]
Yeah, the rushstack eslint patch require() breaks it for me. I wasn't able to make @eslint/eslintrc FlatCompat work because of that line. I even tried converting my config to commonjs, because the docs say `only commonjs supported`, but that seems irrelevant and didn't fix it either. I ended up including eslint plugins react,react-hooks,import,jsx-ally,next myself. And just copy+pasting over the 10 or so rules that eslint-config-next provides. That's really all the config is, the recommended rules from those 5 plugins plus 10 or so modifications. The only other line that is needed is the `eslint-import-resolver-typescript`. All the other settings in there are superfluous or legacy cruft at this point.

https://github.com/vercel/next.js/blob/ca5f29d81b011c3a01e2372118b29836f1df0fa7/packages/eslint-config-next/index.js#L53

Kinda surprised this hasn't been addressed, because the changes needed to update it are so trivial. I guess the engineers at vercel aren't using eslint-config-next in the up-to-date way, that would explain it. Or if there is a general stability-fear-of-change that has taken hold.


Oh, and those 5 plugins mostly don't support flat config yet, but that can easily be worked around with something like
```ts
    plugins: { 'react-hooks': reactHooks },
    rules: reactHooks.configs.recommended.rules,
```
where `reactHooks` refers to the root default export of the plugin. Again, why is this so hard for plugin authors to do? It's so trivial, lol. Apathy? Whatever the explanation it would probably apply to both next and react plugins since several of the core React engineers are employed at Vercel.

**Reply by jdhiro** [reply]
@leerob any chance you could chime in here and let us know when there will be a sensible eslist9/nextjs config available?

**Reply by SouroucheB** [reply]
I'm exactly in the same situation, when I'm running the "next lint" in the terminal, I'm getting the `rushstack` issue.
```
Failed to patch ESLint because the calling module was not recognized.
If you are using a newer ESLint version that may be unsupported, please create a GitHub issue:
https://github.com/microsoft/rushstack/issues
```

Guys, what did you ended to do ? Just rollbacking to Eslint 8 or as @paschun says, copy/past the `eslint-config-next` config ?
TYVM.

**Reply by omermecitoglu** [reply]
> I'm exactly in the same situation, when I'm running the "next lint" in the terminal, I'm getting the `rushstack` issue.
> 
> ```
> Failed to patch ESLint because the calling module was not recognized.
> If you are using a newer ESLint version that may be unsupported, please create a GitHub issue:
> https://github.com/microsoft/rushstack/issues
> ```
> 
> Guys, what did you ended to do ? Just rollbacking to Eslint 8 or as @paschun says, copy/past the `eslint-config-next` config ? TYVM.

https://github.com/vercel/next.js/discussions/49337#discussioncomment-5998603

**Comment by prashantchothani** [comment]
Same for me. The reference to @rushstack/eslint-patch is coming from: node_modules/eslint-config-next/index.js. It contains require('@rushstack/eslint-patch/modern-module-resolution') which is creating an issue for eslint 9 to run. So the twitter post supporting eslint 9 is not correct I feel @leerob

**Reply by leerob** [reply]
It might be worth starting a new project with `create-next-app` and picking ESLint, which will use v9 ([source code](https://github.com/vercel/next.js/blob/canary/packages/create-next-app/templates/default/ts/eslint.config.mjs)), and then working backwards from that example vs. your current repo.

**Reply by prashantchothani** [reply]
This is a new project. However I will try once again as suggested
